rosidl_typesupport_cpp/resource/srv__type_support.cpp.em

```diff
@@ -109,12 +117,86 @@ static const type_support_map_t _@(service.namespaced_type.name)_service_typesup
+(
+    const rosidl_service_introspection_info_t * info,
+    rcutils_allocator_t * allocator,
+    void * request_message,
+    void * response_message,
+    bool enable_message_payload)
+{
+  auto * event_msg = static_cast<@event_type *>(allocator->zero_allocate(1, sizeof(@event_type), allocator->state));
+  if (nullptr == event_msg) {
+    return NULL;
+  }
+  event_msg = new(event_msg) @(event_type)(); 
+
+  event_msg->info.set__event_type(info->event_type);
+  event_msg->info.set__sequence_number(info->sequence_number);
+  event_msg->info.stamp.set__sec(info->stamp_sec);
+  event_msg->info.stamp.set__nanosec(info->stamp_nanosec);
+
+  std::array<uint8_t, 16> client_id;
+  std::move(std::begin(info->client_id), std::end(info->client_id), client_id.begin());
+  event_msg->info.client_id.set__uuid(client_id);
+
+  if (enable_message_payload) {
+    if (nullptr == request_message) {
+      event_msg->response.push_back(*static_cast<@('::'.join([package_name, *interface_path.parents[0].parts, service.namespaced_type.name]))_Response *> (response_message));
+    } else if (nullptr == response_message) {
+      event_msg->request.push_back(*static_cast<@('::'.join([package_name, *interface_path.parents[0].parts, service.namespaced_type.name]))_Request *> (request_message)); }
+      else {
+        // raise an error here?
+    }
+  }
+
+  return event_msg;
+}
+
+#include <memory>
+
+bool
+rosidl_@('_'.join([package_name, *interface_path.parents[0].parts, service.namespaced_type.name]))_introspection_message_destroy(
+void* event_msg,
+rcutils_allocator_t * allocator
+)
+{
+  auto * event_msg_ = static_cast<@event_type *>(event_msg);
+  // event_msg_->@('::'.join([package_name, *interface_path.parents[0].parts]) + '::~' + service.namespaced_type.name + '_Event')();
+  // std::destroy_at(event_msg_); // Why can't I compile with c++17?
+  event_msg_->~@(service.namespaced_type.name)_Event();
+  allocator->deallocate(event_msg, allocator->state);
+  return true;
+}
+
+
 static const rosidl_service_type_support_t @(service.namespaced_type.name)_service_type_support_handle = {
-  ::rosidl_typesupport_cpp::typesupport_identifier,
-  reinterpret_cast<const type_support_map_t *>(&_@(service.namespaced_type.name)_service_typesupport_map),
-  ::rosidl_typesupport_cpp::get_service_typesupport_handle_function,
+  .typesupport_identifier = ::rosidl_typesupport_cpp::typesupport_identifier,
+  .data = reinterpret_cast<const type_support_map_t *>(&_@(service.namespaced_type.name)_service_typesupport_map),
+  .func = ::rosidl_typesupport_cpp::get_service_typesupport_handle_function,
+  .introspection_message_create_handle  = rosidl_@('_'.join([package_name, *interface_path.parents[0].parts, service.namespaced_type.name]))_introspection_message_create,
+  .introspection_message_destroy_handle = rosidl_@('_'.join([package_name, *interface_path.parents[0].parts, service.namespaced_type.name]))_introspection_message_destroy,
```

Comment:
What do you think about using term "event message" instead of "introspection message"? It seems a little more precise (and concise). This would mean renaming these handles (and generated functions accordingly) e.g.

```
event_message_create_handle
event_message_destroy_handle
```

Reply:
Yeah, I agree that it'd be good to rename everything for consistency. This has been resolved in [9e0b9814bcfe332427360b397cdc712fd3aabaf7 ](https://github.com/ros2/rosidl/pull/700/commits/9e0b9814bcfe332427360b397cdc712fd3aabaf7) and 0e03efc8a6f1aa4df8f7b9d023f7eba98a910a3f